### PR TITLE
Added a make_safe_path parameter to paste

### PR DESCRIPTION
**Pull request by laktak**
<del>It allows you to specify 'overwrite', 'rename', or 'rename_ext' (=default, same as the current behaviour) in case of a file name conflict. This obsoletes the overwrite parameter.</del>

It allows you to specify your own function to generate the path for conflicting files.

See https://github.com/ranger/ranger/issues/1616

#### ISSUE TYPE
- Improvement/feature implementation

#### CHECKLIST

- [x] The `CONTRIBUTING` document has been read **[REQUIRED]**
- [x] All changes follow the code style **[REQUIRED]**
- [x] All new and existing tests pass **[REQUIRED]**
- [ ] Changes require config files to be updated
    - [x] Config files have been updated
- [x] Changes require documentation to be updated
    - [x] Documentation has been updated
- [ ] Changes require tests to be updated
    - [ ] Tests have been updated

#### MOTIVATION AND CONTEXT

When you paste a file that already exists (e.g. yy followed by pp), ranger will append a _ to the name.

At the moment this breaks the file extension (e.g. foo.txt becomes foo.txt_).

This PR makes it possible to specify how the conflict should be resolved.

### CODE REVIEW

**Comment by toonn** [comment]
I'd still rather have the more flexible solution I described. Rather than having a long if statement that picks a function based on a string, why not just pass in a function? It's infinitely more useful. I know you can't pass a function from `rc.conf` easily. But I look at this as something you'd customize using `commands.py` instead.

**Comment by laktak** [comment]
I forgot to fix the default arg for CopyLoader but it does take a function as an argument. I described the call chain in the issue and thought that is what you meant.

Can you override actions in `commands.py` or how would you customize it?

File: ranger/core/actions.py [hunk 1]
```diff
@@ -1591,14 +1592,29 @@ def _recurse_hardlinked_tree(self, source_path, target_path):
                 link(source_path,
                      next_available_filename(target_path))
 
-    def paste(self, overwrite=False, append=False, dest=None):
+    def paste(self, conflict='rename_ext', append=False, dest=None, overwrite=False):
```
Comment (toonn) [inline ranger/core/actions.py]: Don't change the argument order. There's no good reason to.

Comment (toonn) [reply ranger/core/actions.py]: I also don't like the string argument, too error prone. Let's just pass a function.

File: ranger/core/actions.py [hunk 2]
```diff
@@ -1591,14 +1592,29 @@ def _recurse_hardlinked_tree(self, source_path, target_path):
...
         """:paste
 
         Paste the selected items into the current directory or to dest
         if provided.
         """
+        # overwrite is obsolete, it was replaced by conflict
+        if overwrite:
+            conflict = 'overwrite'
+
+        if conflict == 'overwrite':
```
Comment (toonn) [inline ranger/core/actions.py]: This whole if statement needs to go, too complicated.

Comment (toonn) [reply ranger/core/actions.py]: This is why we're using a function argument.

File: ranger/core/loader.py [hunk 3]
```diff
@@ -51,12 +52,12 @@ def destroy(self):
 class CopyLoader(Loadable, FileManagerAware):  # pylint: disable=too-many-instance-attributes
     progressbar_supported = True
 
-    def __init__(self, copy_buffer, do_cut=False, overwrite=False, dest=None):
+    def __init__(self, copy_buffer, do_cut=False, resolve_conflict=get_safe_path, dest=None):
```
Comment (toonn) [inline ranger/core/loader.py]: `overwrite`'s not going anywhere. It should take precedence over the new argument.

File: ranger/ext/safe_path.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# This file is part of ranger, the console file manager.
+# License: GNU GPL version 3, see the file "AUTHORS" for details.
+
+import os
+
+APPENDIX = '_'
+
+
+def get_safe_path_classic(dst):
```
Comment (toonn) [inline ranger/ext/safe_path.py]: The original `get_safe_path` is definitely not gonna change names.

File: examples/rc_emacs.conf [hunk 5]
```diff
@@ -335,8 +335,8 @@ map <A-d>  console rename%space
 map <C-e>  eval fm.open_console('rename ' + fm.thisfile.relative_path)
 map <C-a>  eval fm.open_console('rename ' + fm.thisfile.relative_path, position=7)
 
-map <C-y>y  paste
-map <C-y>o  paste overwrite=True
+map <C-y>y  paste conflict=rename_ext
```
Comment (toonn) [inline examples/rc_emacs.conf]: The default bindings are certainly not gonna change.

**Comment by laktak** [comment]
> The intent is good but I feel like we're not getting closer to what I've been suggesting.

It's not like I dislike your suggestions - I did not add a function argument to paste because I have no idea how to configure it. Some help please.

Is this how you would do it? https://github.com/ranger/ranger/blob/master/examples/plugin_new_macro.py

**Comment by toonn** [comment]
No, you'd just write a new command in `commands.py` that's just a thin wrapper over `paste` and passes in a different function for conflict resolution. `commands.py` is 100% a config file from our pov.

**Comment by toonn** [comment]
And don't worry, we'd probably include the "rename_ext" in the default `commands.py` so you won't even actually need to keep track of your own.

**Comment by laktak** [comment]
While looking into this I thought of another approach - how about we simply allow the user to replace the `get_safe_path()` function with his own implementation? E.g. a `set_safe_path_function(...)` that he can call in `commands.py`?

That way we could keep all functions as they are and only need to change the code in `shutil_generatorized.py`.

**Comment by toonn** [comment]
Nope, that's worse. What's wrong with the way I suggested?

**Comment by laktak** [comment]
> No, you'd just write a new command in `commands.py` that's just a thin wrapper over `paste` and passes in a different function for conflict resolution. `commands.py` is 100% a config file from our pov.

Can you tell me how to pass functions from this wrapper over `paste`?

```
class paste_ext(Command):
    def execute(self):
        paste_cls = self.fm.commands.get_command("paste")
        paste = paste_cls("overwrite=False")
        return paste.execute()
```

**Comment by toonn** [comment]
There is no paste "Command", you just call the paste function defined in `actions.py`, you can handle arguments to your custom command if you want but you'd pass those in addition to a new locally defined conflict handling function.

**Comment by laktak** [comment]
Took me a while to see that fm inherits from Actions. I've added `paste_ext` to ` ranger/config/commands.py`.

File: doc/ranger.pod [hunk 6]
```diff
@@ -527,8 +527,8 @@ also "da", "dr" and "dt" shortcuts equivalent to "ya", "yr" and "yt".)
 Paste the files which were previously copied or cut, like pressing Ctrl+V in
 modern GUI programs.
 
-Conflicts will be renamed by appending an _ to the file extension. To change
-this see the paste command.
+Conflicts will be renamed by appending an _ to the full file name. To change
+this and insert a _ before the file extension see commands.py.
```
Comment (toonn) [inline doc/ranger.pod]: Please check elsewhere in the file, I think we use some sort of formatting for files. `F<file.ext>` maybe? Maybe we should also quote the underscores while we're at it? Look for other examples of single literal characters in the manual to check.

File: ranger/config/commands.py [hunk 7]
```diff
@@ -1859,3 +1859,34 @@ def tab(self, tabnum):
...
             if mode
         )
+
+
+class paste_ext(Command):
+    """
+    :paste_ext
+
+    Like paste but renames conflicting files so that the
+    file extension stays intact.
```
Comment (toonn) [inline ranger/config/commands.py]: Maybe admit here that it's best-effort rather than claim it always works : )

File: ranger/core/actions.py [hunk 8]
```diff
@@ -1591,15 +1591,15 @@ def _recurse_hardlinked_tree(self, source_path, target_path):
                 link(source_path,
                      next_available_filename(target_path))
 
-    def paste(self, overwrite=False, append=False, dest=None):
+    def paste(self, overwrite=False, append=False, dest=None, make_safe_path=None):
```
Comment (toonn) [inline ranger/core/actions.py]: Can't we just assign `get_safe_path` by default rather than `None`? Should simplify some of the other code as well.

Comment (laktak) [reply ranger/core/actions.py]: Yes but at the moment `shutil_generatorized.py` appears to be 'private' (it's only imported in the `generate` function of `CopyLoader`). Should we move `get_safe_path` to a new file?

File: ranger/config/commands.py (hunk 7 above)
Comment (laktak) [reply ranger/config/commands.py]: OK

File: doc/ranger.pod (hunk 6 above)
Comment (laktak) [reply doc/ranger.pod]: fixed

**Comment by laktak** [comment]
@toonn ready to merge?

**Comment by toonn** [comment]
I think so, still have to review thoroughly and actually test. Have been busy, will ping you when I get around to it, if necessary.

**Comment by laktak** [comment]
hey @toonn can I help to get this merged? Maybe buy you a coffee? :-)

**Comment by toonn** [comment]
Thanks for the ping, oversight on my part. Don't hold your breath though. Slow going lately.

**Comment by laktak** [comment]
@toonn thank you! Wish you happy holidays! Let me know if you have a donate link (I only found the one for bitcoin which I don't have).

**Comment by toonn** [comment]
We have a bitcoin donation link? That's news to me : )
Afaik the ranger team doesn't currently take donations. Thanks for the gesture anyway!
🎄 Enjoy the festivities!
(There won't be a new ranger under the tree but the new release isn't far off : ))

**Pull request merged**